janus.go

```diff
@@ -251,49 +249,48 @@ func (gateway *Gateway) recv() {
 
 // Info sends an info request to the Gateway.
 // On success, an InfoMsg will be returned and error will be nil.
-func (gateway *Gateway) Info() (*InfoMsg, error) {
-	req, ch := newRequest("info")
-	gateway.send(req, ch)
-
-	msg := <-ch
-	switch msg := msg.(type) {
-	case *InfoMsg:
-		return msg, nil
-	case *ErrorMsg:
-		return nil, msg
-	}
+// func (gateway *Gateway) Info() (*InfoMsg, error) {
+// 	req, ch := newRequest("info")
+// 	gateway.send(req, ch)
 
-	return nil, unexpected("info")
-}
+// 	msg := <-ch
+// 	switch msg := msg.(type) {
+// 	case *InfoMsg:
+// 		return msg, nil
+// 	case *ErrorMsg:
+// 		return nil, msg
+// 	}
+
+// 	return nil, unexpected("info")
+// }
 
 // Create sends a create request to the Gateway.
 // On success, a new Session will be returned and error will be nil.
-func (gateway *Gateway) Create() (*Session, error) {
-	req, ch := newRequest("create")
-	gateway.send(req, ch)
-
-	msg := <-ch
-	var success *SuccessMsg
-	switch msg := msg.(type) {
-	case *SuccessMsg:
-		success = msg
-	case *ErrorMsg:
-		return nil, msg
-	}
-
-	// Create new session
-	session := new(Session)
-	session.gateway = gateway
-	session.ID = success.Data.ID
-	session.Handles = make(map[uint64]*Handle)
-	session.Events = make(chan interface{}, 2)
-
-	// Store this session
-	gateway.Lock()
-	gateway.Sessions[session.ID] = session
-	gateway.Unlock()
+func (gateway *Gateway) CreateSession(ctx context.Context, msg BaseMsg) (*SuccessMsg, error) {
+	ch := make(chan any)
+	gateway.send(msg, msg.ID, ch)
+
+	select {
+	case response := <-ch:
+		switch response := response.(type) {
+		case *SuccessMsg:
+			session := new(Session)
+			session.ID = response.Data.ID
+			session.Handles = make(map[uint64]*Handle)
+			session.gateway = gateway
+			gateway.Lock()
+			gateway.Sessions[session.ID] = session
+			gateway.Unlock()
+			return response, nil
+		case *ErrorMsg:
+			return nil, response
+		default:
+			return nil, nil
 
-	return session, nil
+		}
+	case <-ctx.Done():
```

Comment:
We need to somehow handle the message coming back since it will be ignored by using this approach.